File: frontend/mobile/lib/features/prescriptions/domain/entities/prescription.dart
import 'package:equatable/equatable.dart';

enum PrescriptionStatus { low, medium, high }

class Prescription extends Equatable {
  final String id;
  final String parameter;
  final String value;
  final String recommendation;
  final PrescriptionStatus status;
  final DateTime timestamp;
  final bool isCompleted;
  final String fieldId;
  final String growthStage;
  final double impactScore;
  
  // Additional fields for detailed view
  final String? title;
  final String? category;
  final String? priority;
  final DateTime? dueDate;
  final String? estimatedDuration;
  final List<String>? materials;
  final List<String>? instructions;
  final String? urgency;
  final String? timeline;
  final String? fieldName;
  final String? soilType;

  const Prescription({
    required this.id,
    required this.parameter,
    required this.value,
    required this.recommendation,
    required this.status,
    required this.timestamp,
    this.isCompleted = false,
    required this.fieldId,
    required this.growthStage,
    this.impactScore = 0.0,
    this.title,
    this.category,
    this.priority,
    this.dueDate,
    this.estimatedDuration,
    this.materials,
    this.instructions,
    this.urgency,
    this.timeline,
    this.fieldName,
    this.soilType,
  });

  @override
  List<Object?> get props => [
        id,
        parameter,
        value,
        recommendation,
        status,
        timestamp,
        isCompleted,
        fieldId,
        growthStage,
        impactScore,
        title,
        category,
        priority,
        dueDate,
        estimatedDuration,
        materials,
        instructions,
        urgency,
        timeline,
        fieldName,
        soilType,
      ];

  Prescription copyWith({
    String? id,
    String? parameter,
    String? value,
    String? recommendation,
    PrescriptionStatus? status,
    DateTime? timestamp,
    bool? isCompleted,
    String? fieldId,
    String? growthStage,
    double? impactScore,
    String? title,
    String? category,
    String? priority,
    DateTime? dueDate,
    String? estimatedDuration,
    List<String>? materials,
    List<String>? instructions,
    String? urgency,
    String? timeline,
    String? fieldName,
    String? soilType,
  }) {
    return Prescription(
      id: id ?? this.id,
      parameter: parameter ?? this.parameter,
      value: value ?? this.value,
      recommendation: recommendation ?? this.recommendation,
      status: status ?? this.status,
      timestamp: timestamp ?? this.timestamp,
      isCompleted: isCompleted ?? this.isCompleted,
      fieldId: fieldId ?? this.fieldId,
      growthStage: growthStage ?? this.growthStage,
      impactScore: impactScore ?? this.impactScore,
      title: title ?? this.title,
      category: category ?? this.category,
      priority: priority ?? this.priority,
      dueDate: dueDate ?? this.dueDate,
      estimatedDuration: estimatedDuration ?? this.estimatedDuration,
      materials: materials ?? this.materials,
      instructions: instructions ?? this.instructions,
      urgency: urgency ?? this.urgency,
      timeline: timeline ?? this.timeline,
      fieldName: fieldName ?? this.fieldName,
      soilType: soilType ?? this.soilType,
    );
  }
}

File: backend/src/models/Prescription.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IPrescription extends Document {
  farmId: mongoose.Types.ObjectId;
  title: string;
  description: string;
  priority: 'low' | 'medium' | 'high' | 'urgent';
  status: 'pending' | 'in_progress' | 'completed' | 'cancelled';
  dueDate: Date;
  category: string;
  estimatedDuration: string;
  materials: string[];
  instructions: string[];
  urgency: 'LOW' | 'MEDIUM' | 'HIGH' | 'URGENT';
  timeline: string;
  parameter: string;
  fieldName: string;
  soilType: string;
  growthStage: string;
  completedBy?: mongoose.Types.ObjectId;
  completedAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

const prescriptionSchema = new Schema<IPrescription>({
  farmId: {
    type: Schema.Types.ObjectId,
    required: true,
    ref: 'Farm'
  },
  title: {
    type: String,
    required: true,
    trim: true
  },
  description: {
    type: String,
    required: true,
    trim: true
  },
  priority: {
    type: String,
    enum: ['low', 'medium', 'high', 'urgent'],
    default: 'medium'
  },
  status: {
    type: String,
    enum: ['pending', 'in_progress', 'completed', 'cancelled'],
    default: 'pending'
  },
  dueDate: {
    type: Date,
    required: true
  },
  category: {
    type: String,
    required: true,
    trim: true
  },
  estimatedDuration: {
    type: String,
    required: true,
    trim: true
  },
  materials: [{
    type: String,
    trim: true
  }],
  instructions: [{
    type: String,
    trim: true
  }],
  urgency: {
    type: String,
    enum: ['LOW', 'MEDIUM', 'HIGH', 'URGENT'],
    default: 'MEDIUM'
  },
  timeline: {
    type: String,
    required: true,
    trim: true
  },
  parameter: {
    type: String,
    required: true,
    trim: true
  },
  fieldName: {
    type: String,
    required: true,
    trim: true
  },
  soilType: {
    type: String,
    required: true,
    trim: true
  },
  growthStage: {
    type: String,
    required: true,
    trim: true
  },
  completedBy: {
    type: Schema.Types.ObjectId,
    ref: 'User'
  },
  completedAt: {
    type: Date
  }
}, {
  timestamps: true
});

// Create indexes for better query performance
prescriptionSchema.index({ farmId: 1, status: 1 });
prescriptionSchema.index({ dueDate: 1 });
prescriptionSchema.index({ priority: 1 });
prescriptionSchema.index({ createdAt: -1 });

const Prescription = mongoose.model<IPrescription>('Prescription', prescriptionSchema);

export default Prescription;